---
title: "Modeling the origin of mutations identified in a liquid biopsy: cancer or clonal hematopoiesis?"
author: "Adith S. Arun and Robert B. Scharpf"
date: "`r format(Sys.Date())`"
output: BiocStyle::html_document
abstract: >
  Mutation-based approaches for detection of cancer from cell-free DNA (cfDNA) using liquid biopsies have the potential to track a patient's response to treatment, enabling effective and timely decisions on therapy.  However, mutations arising from clonal hematopoeisis (CH) are common and tumor biopsies for definitive identification of the origin of these mutations is not always available. Sequencing of matched cells from buffy coat and the presence of mutations in these cells has been used as a test to rule-in CH, but uneven sequencing depths between matched cell-free DNA and buffy coat samples and the potential for contamination of buffy coat with circulating tumor cells (CTCs) are not captured by rule-based analyses.  This package provides estimates of the marginal likelihood for CH- and tumor-origin models of cfDNA mutations in plasma, requiring sequencing of cfDNA and matched buffy coat and the allele frequencies of high quality alignments available from standard mutation callers.
vignette: >
  %\VignetteIndexEntry{Modeling the origin of mutations in a liquid biopsy: cancer or clonal hematopoiesis?}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

Suppose we identified 4 fragments with a variant out of 1000 cell-free DNA (cfDNA) fragments in plasma, and 1 fragment with the same mutation in sequenced white blood cells.
The possible scenarios that explain the observed data are:
(1) the variant is tumor-derived and some CTCs are mixed in with the WBCs,
(2) the variant is derived from clonal hematopoeisis of indeterminant potential (CHIP).
A third explanation is sequencing error, but after removing PCR duplicates the probability of observing the same event in plasma and WBCs should be negligible relative to the scenarios listed above.
In the absence of additional information, our prior for the prevalence of CTCs inter-mixed with WBCs will be influential on how we interpret this data.

# Data organization

```{r setup, message=FALSE}
library(magrittr)
library(tidyverse)
library(ggrepel)
library(plasmut)
library(qpdf)
knitr::opts_chunk$set(cache=TRUE)

lo <- function(p) log(p/(1-p))
```

We assume the following minimal representation for the mutation data such that each row uniquely identifies a mutation within a sample.
As inference for a mutation is independent of other mutations in the same sample, the Bayesian model is implemented independently for each row in the table.

```{r data_wrangling}
## sample data
datA <- tibble(y=c(4, 1),
              n=c(1000, 1000),
              analyte=c("plasma", "buffy coat"),
              mutation="mutA",
              sample_id="id1")
datB <- tibble(y=c(4, 0),
              n=c(1000, 1000),
              analyte=c("plasma", "buffy coat"),
              mutation="mutB",
              sample_id="id1")
dat <- bind_rows(datA, datB) %>%
    unite("uid", c(sample_id, mutation), remove=FALSE) %>%
    group_by(uid) %>%
    nest()

## required format for plasmut
dat
```


# Approach and implementation

## Bayesian model

From matched sequencing, for a given mutation, we are given the number of mutant reads and distinct reads in plasma ($y_p$, $n_p$); the number of mutant reads and distinct reads in the buffy coat layer ($y_w$, $n_w$). Using this data, we want to estimate the posterior odds that the cell-free DNA mutation is tumor-derived. If this probability is low, the mutation may be a product of clonal hematopoiesis (CH). The posterior odds is equal to the Bayes factor times the prior odds of the mutation being tumor-derived: $\frac{p(S | y_p, y_w, n_p, n_w)}{p(H | y_p, y_w, n_p, n_w)} = \frac{p(y_w, y_p | n_p, n_w, S)}{p(y_w, y_p | n_p, n_w, H)} \cdot \frac{p(S)}{p(H)}$. 

To compute the Bayes factor we must estimate the marginal likelihood when we assume the mutation is tumor-derived and CH. There exists a true but unobserved mutant allele fraction (MAF) in plasma and white blood cells, $\theta_p$ and $\theta_w$ respectively, and to estimate the marginal likelihoods, we integrate out these parameters.  When we assume the mutation is a product of CH (model H), we estimate $p(y_p, y_w | n_p, n_w, H) = \int_{\theta} p(y_p | n_p, \theta)$. This single integral is possible becuase given that the mutation is a result of CH, it is reasonable to assume that $\theta_p = \theta_w$. The prior for $\theta$ is a diffuse Beta distribution to support the possibility of both rare and common CH mutations. When we assume the mutation is tumor-derived or somatic (model S), we estimate $p(y_p, y_w | n_p, n_w, S) = \int_{\theta_p} p(y_p | n_p, \theta_p, S) d\theta_w\int_{\theta_w} p(y_w | n_w, \theta_w, S) d\theta_p$ since the double integral can be factored into two single integrals since $\theta_p$ is independent of $\theta_w$. A diffuse prior for $\theta_p$ is chosen to allow for many different somatic mutation possibilities. There are two possibilities for a prior on $\theta_w$. If we believe that there are no circulating tumor cells (CTCs) within the buffy coat layer, then $\theta_w = 0$. However, if we believe that CTCs exist within the buffy coat layer, a prior distribution near zero would be appropriate (i.e., Beta(1, $10^3$)).      

We can compute marginal likelihoods for model S and H by simply drawing Monte Carlo samples from the $\theta$ priors we defined. But, when the likelihood is more concentrated than the prior, this sampling approach can be inefficient and numerically unstable. To solve these issues, we implement an importance sampling approach. 

## Importance sampling

The general idea for importance sampling is that we want to sample $\theta$'s from a density that has a similar shape as the posterior but fatter tails so that we avoid spending time on simulations that have very low likelihood. Importance sampling should make estimation of the marginal likelihood more efficient, and the resulting Bayes factor estimates more stable with fewer Monte Carlo samples.

## Model assumptions and priors
Below, we assume that
(1) the probability that CTCs are mixed in with the WBCs is small (e.g., 1 CTC per 1,000 cells) using a Beta(1, 10^3) prior,
(2) the prior for the MAF of ctDNA in plasma is relatively diffuse (Beta(1, 9)), and
(3) the prior for germline or CHIP variants in WBCs is also relatively diffuse (Beta(1, 9)).

Our importance sampler will be a mix of the prior and posterior with a default weight of 0.1 on the prior, resulting in a fatter tailed distribution than the posterior.

```{r parameters}
## Parameters
param.list <- list(ctc=list(a=1, b=10e3),
                   ctdna=list(a=1, b=9),
                   chip=list(a=1, b=9),
                   montecarlo.samples=50e3,
                   prior.weight=0.1)
```

## Monte Carlo estimation of the marginal likelihood and Bayes factors

Below, we estimate the marginal likelihood for the mutation frequencies under two models:  (1) a somatic model that assumes that any mutations in plasma and buffy coat are tumor derived, and (2) a germline or CHIP model that assumes that the mutations observed in plasma and buffy coat are not tumor-derived.  All Monte Carlo samples are saved by default.  For running this model on large datasets, we recommend saving only the marginal likelihoods and Bayes factors by setting  `save_montecarlo=FALSE`.

```{r montecarlo}
stats <- importance_sampler(dat$data[[1]], param.list)
stats  ## with all the monte carlo samples
## Just the mutation-level summary statistics (marginal likelihoods and bayes factors)
importance_sampler(dat$data[[1]], param.list, save_montecarlo=FALSE)
```

With our prior that CTCs are rare, we view the plasma MAF of 4/1000 and buffy coat MAF of 1/1000 as very weak evidence that the mutation is tumor derived (log Bayes factor `round(stats$bayesfactor$bayesfactor, 2)`).  Had we instead observed no mutations in buffy coat, the log Bayes factor increases by 3 and we would conclude that there is strong evidence in support of a tumor-derived mutation.

```{r scenariob}
importance_sampler(dat$data[[2]], param.list, save_montecarlo=FALSE)
```

Note, that the log marginal likelihood of the CTC model is higher as observing no mutations in 1000 reads from buffy coat is more consistent with our prior; and the log marginal likelihood of the chip model decreases as we would anticipate observing mutant reads in the buffy coat if the mutation was derived from CHIP.

## Efficiency of importance sampler

As long as `montecarlo.samples` is big enough, we should obtain a similar estimate of the marginal likelihood without importance sampling.
Since our target distribution $g$ is a mixture of the prior and posterior with weight `prior.weight`, setting `prior.weight=1` just samples $\theta$'s from our prior (i.e., importance sampling is not implemented).  Below, we compare the stability of the Bayes factor estimate as a function of the Monte Carlo sample size and prior weight:

```{r prior.weight, cache=TRUE}
fun <- function(montecarlo.samples, data,
                param.list, prior.weight=0.1){
    param.list$montecarlo.samples <- montecarlo.samples
    param.list$prior.weight <- prior.weight
    res <- importance_sampler(data, param.list,
                              save_montecarlo=FALSE) %>%
        mutate(montecarlo.samples=montecarlo.samples,
               prior.weight=param.list$prior.weight)
    res
}
fun2 <- function(montecarlo.samples, data,
                 param.list, prior.weight=0.1,
                 nreps=100){
    res <- replicate(nreps, fun(montecarlo.samples, data,
                                param.list,
                                prior.weight=prior.weight),
                     simplify=FALSE) %>%
        do.call(bind_rows, .) %>%
        group_by(montecarlo.samples, prior.weight) %>%
        summarize(mean_bf=mean(bayesfactor),
                  sd_bf=sd(bayesfactor),
                  .groups="drop")
    res
}
S <- c(100, 1000, seq(10e3, 50e3, by=10000))
results <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list)
results2 <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list,
            prior.weight=1)
combined <- bind_rows(results, results2)
```


```{r standardev, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    ggplot(aes(montecarlo.samples, sd_bf)) +
    geom_point(aes(color=prior.weight)) +
    geom_line(aes(group=prior.weight, color=prior.weight)) +
    scale_y_log10() +
    theme_bw(base_size=16) +
    xlab("Monte Carlo samples") +
    ylab("Standard deviation of\n log Bayes Factor")
```

Note that with importance sampling, relatively stable estimates for the Bayes factor are obtained with as few as 10,000 Monte Carlo samples while sampling from the prior distribution is very unstable for small sample sizes.

```{r means, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    filter(montecarlo.samples > 100) %>%
    ggplot(aes(prior.weight, mean_bf)) +
    geom_point() +
    theme_bw(base_size=16) +
    ylab("Mean log Bayes factor") +
    xlab("Prior weight")
```


# Application to van't Erve et al.

We apply our methods to the dataset introduced in [van't Erve et al](https://aacrjournals.org/clincancerres/article/29/5/899/716607/Metastatic-Colorectal-Cancer-Treatment-Response). The cell-free DNA and buffy coat were sequenced for patients with metastatic colorectal cancer and we can estimate Bayes factors for these mutations. 

We select four mutations and run the importance sampler for these candidate mutations independently. 

```{r vanterve}
data(crcseq)

crcseq %>% select(-position)
```

```{r vanterve importance sampling}
params <- list(ctdna = list(a = 1, b = 9), 
               ctc = list(a = 1, b = 10^3), 
               chip = list(a= 1, b = 9), 
               montecarlo.samples = 50e3, 
               prior.weight = 0.1)

muts <- unite(crcseq, "uid", c(patient, gene), remove = FALSE) %>% 
        group_by(uid) %>% nest()

#Each element of the data column contains a table with the variant and total allele counts in plasma and buffy coat. 

#Run the importance sampler
muts$IS <- muts$data %>% map(importance_sampler, params)

bf <- apply(muts, 1, function(x) x$data %>% select(-position) %>% mutate(bayes_factor = x$IS$bayesfactor$bayesfactor)) 

bf %>% do.call(rbind, .) %>% as_tibble() %>% select(patient, gene, aa, bayes_factor) %>% distinct()
```

We report the log Bayes factor for the four mutations. The Bayes factor informs us of the likelihood that a mutation is tumor-derived. The log Bayes factor is equal to the log posterior odds when the prior odds of a mutation being tumor-derived are 1, which is a good uninformative assumption in the majority of cases. Let us examine the E1306* APC mutation for patient 12. When sequencing the buffy coat, we observe 0 variant reads across 963 total reads. But, when sequencing the cell-free DNA, we observed 395 variant reads across 1750 total reads. Intuitively, this means that the mutation is not found in white blood cells but is present in cell-free DNA fragments meaning that the source of the mutation must be from a non-WBC source. We expect the mutation to be a product of the tumor whose cellular turnover shed mutant DNA fragments into the bloodstream. Concordantly, the log Bayes factor is 190 which translates to a probability of being tumor derived of $\frac{exp(190)}{exp(190) + 1}$ or 1. 

On the other hand, we observe 5 mutant reads across 1495 total WBC reads for the M237K TP53 mutation; 15 mutant reads across 2969 total cfDNA reads. The observed mutant read rate is roughly equal in WBC and cfDNA, and we just happen to observe more cfDNA reads in general. This suggests that the variant is likely not tumor derived and probably results from CH. There is of course the possibility that there is an abnormally high CTC fraction in WBC harboring this mutation but this is not likely. We would expect the probability of the mutation being tumor derived to be low. As expected, this probability is estimated to be $\frac{exp(-1.1)}{exp(-1.1) + 1}$ or 0.25. 

# Session information

```{r session}
sessionInfo()
```


















