---
title: "Walkthrough of plasmut package use for matched white blood cell and cell-free DNA mutants sequenced"
author: "Adith S. Arun and Robert B. Scharpf"
date: "`r format(Sys.Date())`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{plasmut}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
An example workflow of the plasmut package is shown using data from the CAIRO5 study (van't Erve and Medina et. al 2022). For a group of patients with cancer, both cell-free DNA and white blood cell DNA were sequenced. The distinct and total reads for each mutation present in cell-free DNA are displayed in the data. Corresponding white blood cell total and distinct reads are displayed as well.

From this information, we wish to assign labels to each of the mutations. Specifically, for mutations seen in cell-free DNA and not in white blood cells (distinct reads wbc = 0), we estimate the probability that the mutation arises from the tumor. This is done in a Bayesian framework and approximates integrals as Monte Carlo sums. Please see the paper for a workthrough of the math underlying this. Additionally, for mutations seen in white blood cells and cell-free DNA, we classify them as either hematopoietic or germline variants in an unsupervised manner.

```{r setup, message=FALSE}
library(plasmut)
library(tidyverse)
library(readxl)
```

Read in the data and convert columns to their natural types.

```{r load data}
extdir <- system.file("extdata", package="plasmut")
fpath <- file.path(extdir, "cairo5-matched-sequencing.csv")
df <- read_csv(fpath) %>% type.convert(as.is=TRUE)
```

Call the plasmut function posterior.odds() and posterior.probability() that computes the probability of a mutation being tumor derived.

@Commments:
- before computing the posterior odds, we should show how to derive the bayes factor
- the posterior odds does not require a function.  It is just the bayes factor times the prior odds.
- how is `cols` checked? Should we require that this be a named vector with names that map to pdmr, pdr, ...?
- each function we export should have a working example
- should we adopt the convention of `_` instead of `.` in function names?

```{r compute probability of tumor-derived mutation}
cols <- c("cfDNA distinct mutant reads",
          "cfDNA distinct reads",
          "wbc distinct mutant reads",
          "wbc distinct reads")
podds <- posterior.odds(data=df,
                        cols=cols,
                        prior.odds=1) %>%
    posterior.probability()
```

Call the plasmut function `stratify_wbc_vars` that identifies which white blood cell mutants are germline versus hematopoietic.

```{r stratify wbc variants}
wbc.var.type <- stratify_wbc_vars(data=df,
                                  cols=cols)
```

Compute mutant allele fractions (MAFs) and clean data.

```{r data cleaning and organization}
df <- df %>%
    mutate(prob.tumor.specific=podds, ##@This argument should be post_odds or similar as it is an odds and not a probability
           wbc.var.type=wbc.var.type) %>%
    mutate(wbcmaf=`wbc distinct mutant reads`/`wbc distinct reads`,
           cfdnamaf=`cfDNA distinct mutant reads`/`cfDNA distinct reads`) %>%
    mutate(cfdnamaf=ifelse(is.nan(cfdnamaf), 0, cfdnamaf))

df <- df %>% filter(g_include == "yes")
```

We estimate the probability of a mutation being somatic (tumor-derived) using only information from cell-free DNA and white blood cells with posterior.odds() and posterior.probability(). We can assess the accuracy of this method by sequencing the tumor and observing whether the mutation is found there. We sequenced 42 cases where the tumor displayed the same variant as the cell-free DNA. Our estimation procedure assigned a probability of tumor of 1 to 41 of these cases and a probability of 0.987 to the other, which was consistent with the tumor sequencing results.

```{r estimation accuracy}
df %>% select(cfdnamaf, wbcmaf, prob.tumor.specific, type) %>%
    filter(type=="Tumor-confirmed") %>%
    ## @ again, this is an odds and not a probability
    mutate(prob.tumor.specific = round(prob.tumor.specific, 3)) %>%
    group_by(prob.tumor.specific) %>%
    summarize(category=unique(type), n=n())
```

Finally, we can visualize the results of this approach. The final object here is a ggplot object that can be further customized.
```{r visualization, fig.align="center", fig.width=7, fig.height=7}
cfvar_plot <- geom_cfvar(df)
cfvar_plot + theme(axis.title=element_text(size=16),
                   axis.text=element_text(size=13),
                   legend.text=element_text(size=12),
                   legend.title=element_text(size=14))
```
