---
title: "Walkthrough of plasmut package use for matched white blood cell and cell-free DNA mutants sequenced"
author: "Adith S. Arun and Robert B. Scharpf"
date: "`r format(Sys.Date())`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{plasmut}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
An example workflow of the plasmut package is shown using data from the CAIRO5 study (van't Erve and Medina et. al 2022). For a group of patients with cancer, both cell-free DNA and white blood cell DNA were sequenced. The distinct and total reads for each mutation present in cell-free DNA are displayed in the data. Corresponding white blood cell total and distinct reads are displayed as well. 

From this information, we wish to assign labels to each of the mutations. Specifically, for mutations seen in cell-free DNA and not in white blood cells (distinct reads wbc = 0), we estimate the probability that the mutation arises from the tumor. This is done in a Bayesian framework and approximates integrals as Monte Carlo sums. Please see the paper for a workthrough of the math underlying this. Additionally, for mutations seen in white blood cells and cell-free DNA, we classify them as either hematopoietic or germline variants in an unsupervised manner.

```{r setup, message=FALSE}
library(plasmut)
library(tidyverse)
library(readxl)
```

Read in the data and convert columns to their natural types. 
```{r load data}
extdir <- system.file("extdata", package="plasmut")
fpath <- file.path(extdir, "cairo5-matched-sequencing.csv")

df <- read_csv(fpath, show_col_types=FALSE) %>% type.convert(as.is=TRUE)

df
```

Call the plasmut function bayes.factor() to compute bayes factors for the probability that a mutation is tumor derived. 
```{r compute bayes factors}
bayesfactors <- bayes.factor(data=df, cols=c("cfDNA distinct mutant reads","cfDNA distinct reads","wbc distinct mutant reads","wbc distinct reads")) 
```

posterior.odds() multiplies each bayes factor to its prior. It is assumed, unless specified, that the prior is 1 indicating an equal prior possibility of each mutation being hematopoeitic or tumor derived. The second, optional argument to posterior.odds() is a vector corresponding to the prior for each mutation being considered. 

posterior.probability() computes the probability of a mutation being tumor derived. 
```{r compute probability of tumor specific alteration}
posteriorodds <- posterior.odds(bayesfactors)
prob <- posterior.probability(posteriorodds)
```


Call the plasmut function stratify_wbc_vars that identifies which white blood cell mutants are germline versus hematopoietic. 
```{r stratify wbc variants}
wbc.var.type <- stratify_wbc_vars(data=df, cols=c("cfDNA distinct mutant reads","cfDNA distinct reads","wbc distinct mutant reads","wbc distinct reads"))
```

Compute mutant allele fractions (MAFs) and clean data. The first few alterations' bayes factors and somatic probability are shown. 

```{r data cleaning and organization}
df <- df %>% mutate(prob.tumor.specific=prob, wbc.var.type=wbc.var.type, bayes.factor=bayesfactors) %>% mutate(wbcmaf=`wbc distinct mutant reads`/`wbc distinct reads`, cfdnamaf=`cfDNA distinct mutant reads`/`cfDNA distinct reads`) %>% mutate(cfdnamaf=ifelse(is.nan(cfdnamaf), 0, cfdnamaf))

df <- df %>% filter(g_include == "yes")

df %>% relocate(Gene, Patient, prob.tumor.specific, bayes.factor, cfdnamaf, wbcmaf) %>% head()
```

We estimate the probability of a mutation being somatic (tumor-derived) using only information from cell-free DNA and white blood cells with posterior.odds() and posterior.probability(). We can assess the accuracy of this method by sequencing the tumor and observing whether the mutation is found there. We sequenced 42 cases where the tumor displayed the same variant as the cell-free DNA. Our estimation procedure assigned a probability of tumor of 1 to 41 of these cases and a probability of 0.987 to the other, which was consistent with the tumor sequencing results. 

```{r estimation accuracy}
df %>% select(cfdnamaf, wbcmaf, prob.tumor.specific, type) %>% filter(type=="Tumor-confirmed") %>% mutate(prob.tumor.specific = round(prob.tumor.specific, 3)) %>% group_by(prob.tumor.specific) %>% summarize(category=unique(type), n=n())
```

Finally, we can visualize the results of this approach
```{r visualization, fig.align="center", fig.width=9, fig.height=9, warning=FALSE}
cluster.annotate <- df %>% select(wbc.var.type, cfdnamaf, wbcmaf) %>% group_by(wbc.var.type) %>% summarize(wbclevel=mean(wbcmaf)) %>% arrange(desc(wbclevel)) %>% filter(!is.na(wbc.var.type))

df <- df %>% mutate(status = wbc.var.type) %>% mutate(status = ifelse(is.na(status), "Likely tumor-specific", wbc.var.type)) %>% mutate(status = ifelse(status==cluster.annotate$wbc.var.type[1], "Germline variants", status)) %>% mutate(status = ifelse(status==cluster.annotate$wbc.var.type[2], "Hematopoietic variants", status)) %>% mutate(prob=status) %>% mutate(prob=ifelse(status=="Likely tumor-specific", prob.tumor.specific, prob)) %>% mutate(prob=ifelse(status=="Germline variants", 2, prob), prob=ifelse(status=="Hematopoietic variants", 0, prob)) %>% mutate(prob=as.numeric(prob))

df <- df %>% mutate(shapestatus=ifelse(type=="Tumor-confirmed", "Tumor confirmed", "Tumor not sequenced"))

cfvar_plot <- ggplot(df, aes(x=wbcmaf, y=cfdnamaf)) +
    geom_jitter(aes(color = prob, fill = status, shape=shapestatus),
                fill = "black", stat = "identity", lwd = 3, width = 0.1, alpha = 0.7) +
    scale_color_gradientn(colors = c("red3", "orange", "blue", "black", "black", "orange"), 
                          na.value = "transparent",
                          breaks=c(0.6, 0.8, 1, 2, 0),
                          labels=c(0.6, 0.8, 1,"Germline variants", "Hematopoietic variants"),
                          limits=c(0 , 3)) +
    scale_x_continuous(labels = scales::percent_format(accuracy = 0.1),
                       breaks=c(0.000084, 0.001, 0.01,  0.1, 0.5),
                       trans='log10',
                       oob = scales::squish_infinite) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 0.1),
                       breaks=c(0.000084, 0.001, 0.01,  0.1, 1),
                       trans='log10',
                       oob = scales::squish_infinite) +
    scale_shape_manual(values = c(17, 16)) + 
    labs(x="White blood cell MAF (%)",
          y="cfDNA MAF (%)")+
    theme_bw() + 
    theme(
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      plot.tag = element_text()) +
    theme(axis.line = element_line(color = 'black')) +
    theme(axis.text.x = element_text(vjust=1.2, hjust= 0.3)) +
    theme(legend.direction = "vertical", legend.box = "horizontal",
          legend.background = element_rect(fill = "transparent"),
          legend.position = c(0.47, 0.88),
          legend.key = element_rect(size = 5), #legend.box.background = element_rect(colour = "black"),
          legend.key.size = unit(0.4, "cm")) +
    guides(color=guide_legend(title="Probability\nTumor-Specific", nrow = 3),
           fill = guide_legend(title=""),
           shape = guide_legend(title="")) +
    geom_vline(xintercept=c(0.00015), linetype="dashed", size = 0.2) +
    geom_vline(xintercept=c(0.00009), linetype="dashed", size = 0.2, color = NA) +
    geom_vline(xintercept=c(0.25), linetype="dashed", size = 0.2) +
    geom_abline(slope=1, intercept=0, color="gray", alpha = 0.4, size = 0.4)

cfvar_plot + theme(axis.title=element_text(size=16), axis.text=element_text(size=13), legend.text=element_text(size=12), legend.title=element_text(size=14))
```
There are a number of mutations with a probability of being tumor specific below 0.9. We can notice that although the white blood cell MAF is zero for these mutations, the number of cell-free DNA mutant reads is very low relative to the total reads. This naturally, introduces some doubt into our minds as to whether the mutant reads are tumor specific, or arising from a different process / tissue. This uncertainty is quantified in our method with reduced probability estimates, something a binary "tumor or not" approach does not capture. 
```{r lower probability tumor specific mutations}
df %>% filter(wbcmaf == 0) %>% arrange(prob) %>% select(prob, `cfDNA distinct mutant reads`, `cfDNA distinct reads`, `wbc distinct mutant reads`, `wbc distinct reads`, type, Gene, Patient) %>% slice(1:10)
```

```{r bayes factor relationship to cfDNA}
cor.val <- df %>% select(prob.tumor.specific, bayes.factor, cfdnamaf, wbcmaf, Gene, Patient) %>% mutate(logbf=log(bayes.factor)) %>% filter(!is.infinite(logbf)) %>% select(cfdnamaf, logbf) %>% cor(method="spearman") %>% as_tibble() %>% slice(1) %>% pull(logbf) %>% round(2)

corr <- paste0("Corr: ", cor.val)

plt <- df %>% select(prob.tumor.specific, bayes.factor, cfdnamaf, wbcmaf, Gene, Patient) %>% mutate(logbf=log(bayes.factor)) %>% filter(!is.infinite(logbf)) %>% ggplot(aes(x=logbf, y=cfdnamaf)) + geom_point(size=1.5) + theme_bw() + ylab("cfDNA MAF") + xlab("log(Bayes Factor)") + theme(panel.grid.major=element_blank(), panel.grid.minor=element_blank()) + geom_smooth(method="loess", se = F) + geom_text(x=400, y=0.22, label=corr, size=4)
```


```{r tile plot, height=6, width=10}

#gene by patient matrix
#each tile is the mutation: 
  # colored by whether it is hematopoeitic / germline / somatic
      
  # colored by if bayes factor is above a certain level 
    # this level is such that the probability of somatic is greater than 0.9 
        # df %>% filter(prob.tumor.specific > 0.9) %>% pull(bayes.factor) %>% min() --> 18.01


plt.df <- df %>% select(status, Gene, Patient, bayes.factor) %>% mutate(border=ifelse(bayes.factor > 18.01, "> 18.01", "Else")) %>% select(-bayes.factor) %>% mutate(alt.source=ifelse(status=="Likely tumor-specific", "Tumor", status)) %>% select(-status)

ordered.levels <- plt.df %>% group_by(Gene) %>% summarize(n=n()) %>% arrange(desc(n)) %>% pull(Gene)

gene.numbers <- plt.df %>% select(Gene) %>% distinct() %>% mutate(gene.num=1:n())

plt.df <- plt.df %>% left_join(gene.numbers, by=c("Gene"="Gene"))

fig <- plt.df %>% ggplot(aes(y=factor(Gene, levels=ordered.levels), x=factor(Patient), fill=alt.source, color=border)) + geom_tile(size=1.25, alpha=0.7, width=0.85, height=0.75) + theme_minimal() + ylab("Gene") + xlab("Patient") + theme(panel.grid.major=element_blank()) + theme(axis.text.x=element_blank()) + scale_color_manual(values=c("firebrick", "darkblue"), name="Bayes Factor") + scale_fill_manual(values=c("mediumorchid", "cornflowerblue", "sienna"), name="Alteration\nSource")

#10x6
fig
```


```{r sample columns of data frame to show for paper}
df %>% select(Gene, Patient, `cfDNA distinct mutant reads`, `cfDNA distinct reads`, `wbc distinct reads`, `wbc distinct mutant reads`) %>% slice(1:4)
```

