---
title: "Modeling the origin of mutations identified in a liquid biopsy: cancer or clonal hematopoiesis?"
author: "Adith S. Arun and Robert B. Scharpf"
date: "`r format(Sys.Date())`"
output: BiocStyle::html_document
abstract: >
  Mutation-based approaches for detection of cancer from cell-free DNA (cfDNA) using liquid biopsies have the potential to track a patient's response to treatment, enabling effective and timely decisions on therapy.  However, mutations arising from clonal hematopoeisis (CH) are common and tumor biopsies for definitive identification of the origin of these mutations is not always available. Sequencing of matched cells from buffy coat and the presence of mutations in these cells has been used as a test to rule-in CH, but uneven sequencing depths between matched cell-free DNA and buffy coat samples and the potential for contamination of buffy coat with circulating tumor cells (CTCs) are not captured by rule-based analyses.  This package provides estimates of the marginal likelihood for CH and cancer tissue-of-origin models of cell-free DNA mutations in plasma, requiring only the allele frequencies of high quality alignments available from standard mutation callers.
vignette: >
  %\VignetteIndexEntry{Modeling the origin of mutations in a liquid biopsy: cancer or clonal hematopoiesis?}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

Suppose we identified 4 fragments with a variant out of 1000 cell-free DNA (cfDNA) fragments in plasma, and 1 fragment with the same mutation in sequenced white blood cells.
The possible scenarios that explain the observed data are:
(1) the variant is tumor-derived and some CTCs are mixed in with the WBCs,
(2) the variant is derived from clonal hematopoeisis of indeterminant potential (CHIP).
A third explanation is sequencing error, but after removing PCR duplicates the probability of observing the same event in plasma and WBCs should be negligible relative to the scenarios listed above.
In the absence of additional information, our prior for the prevalence of CTCs inter-mixed with WBCs will be influential on how we interpret this data.

# Data organization

```{r setup, message=FALSE}
library(magrittr)
library(tidyverse)
library(ggrepel)
library(plasmut)
library(qpdf)
knitr::opts_chunk$set(cache=TRUE)
lo <- function(p) log(p/(1-p))
```

We assume the following minimal representation for the mutation data such that each row uniquely identifies a mutation within a sample.
As inference for a mutation is independent of other mutations in the same sample, the Bayesian model is implemented independently for each row in the table.

```{r data_wrangling}
## Data
datA <- tibble(y=c(4, 1),
              n=c(1000, 1000),
              analyte=c("plasma", "buffy coat"),
              mutation="mutA",
              sample_id="id1")
datB <- tibble(y=c(4, 0),
              n=c(1000, 1000),
              analyte=c("plasma", "buffy coat"),
              mutation="mutB",
              sample_id="id1")
dat <- bind_rows(datA, datB) %>%
    unite("uid", c(sample_id, mutation), remove=FALSE) %>%
    group_by(uid) %>%
    nest()
## Required format
dat
```


# Approach and implementation

The general idea for importance sampling is that we want to sample $\theta$'s from a density that has a similar shape as the posterior but fatter tails so that we avoid spending time on simulations that have very low likelihood. Importance sampling should make estimation of the marginal likelihood more efficient, and the resulting Bayes factor estimates more stable with fewer Monte Carlo samples.

## Model assumptions and priors
Below, we assume that
(1) the probability that CTCs are mixed in with the WBCs is small (e.g., 1 CTC per 10,000 WBCs) using a Beta(1, 9999) prior,
(2) the prior for the MAF of ctDNA in plasma is relatively diffuse (Beta(1, 9)), and
(3) the prior for germline or CHIP variants in WBCs is also relatively diffuse (Beta(1, 9)).

Our importance sampler will be a mix of the prior and posterior with a default weight of 0.1 on the prior, resulting in a fatter tailed distribution than the posterior.

```{r parameters}
## Parameters
param.list <- list(ctc=list(a=1, b=10e3),
                   ctdna=list(a=1, b=9),
                   chip=list(a=1, b=9),
                   montecarlo.samples=50e3,
                   prior.weight=0.1)
```

## Monte Carlo estimation of the marginal likelihood and Bayes factors

Below, we estimate the marginal likelihood for the mutation frequencies under two models:  (1) a somatic model that assumes that any mutations in plasma and buffy coat are tumor derived, and (2) a germline or CHIP model that assumes that the mutations observed in plasma and buffy coat are not tumor-derived.  All Monte Carlo samples are saved by default.  For running this model on large datasets, we recommend saving only the marginal likelihoods and Bayes factors by setting  `save_montecarlo=FALSE`.

```{r montecarlo}
stats <- importance_sampler(dat$data[[1]], param.list)
stats  ## with all the monte carlo samples
## Just the mutation-level summary statistics (marginal likelihoods and bayes factors)
importance_sampler(dat$data[[1]], param.list, save_montecarlo=FALSE)
```

With our prior that CTCs are rare, we view the plasma MAF of 4/1000 and buffy coat MAF of 1/1000 as very weak evidence that the mutation is tumor derived (log Bayes factor `round(stats$bayesfactor$bayesfactor, 2)`).  Had we instead observed no mutations in buffy coat, the log Bayes factor increases by 3 and we would conclude that there is strong evidence in support of a tumor-derived mutation.

```{r scenariob}
importance_sampler(dat$data[[2]], param.list, save_montecarlo=FALSE)
```

Note, that the log marginal likelihood of the CTC model is higher as observing no mutations in 1000 reads from buffy coat is more consistent with our prior; and the log marginal likelihood of the chip model decreases as we would anticipate observing mutant reads in the buffy coat if the mutation was derived from CHIP.

## Efficiency of importance sampler

As long as `montecarlo.samples` is big enough, we should obtain a similar estimate of the marginal likelihood without importance sampling.
Since our target distribution $g$ is a mixture of the prior and posterior with weight `prior.weight`, setting `prior.weight=1` just samples $\theta$'s from our prior (i.e., importance sampling is not implemented).  Below, we compare the stability of the Bayes factor estimate as a function of the Monte Carlo sample size and prior weight:

```{r prior.weight, cache=TRUE}
fun <- function(montecarlo.samples, data,
                param.list, prior.weight=0.1){
    param.list$montecarlo.samples <- montecarlo.samples
    param.list$prior.weight <- prior.weight
    res <- importance_sampler(data, param.list,
                              save_montecarlo=FALSE) %>%
        mutate(montecarlo.samples=montecarlo.samples,
               prior.weight=param.list$prior.weight)
    res
}
fun2 <- function(montecarlo.samples, data,
                 param.list, prior.weight=0.1,
                 nreps=100){
    res <- replicate(nreps, fun(montecarlo.samples, data,
                                param.list,
                                prior.weight=prior.weight),
                     simplify=FALSE) %>%
        do.call(bind_rows, .) %>%
        group_by(montecarlo.samples, prior.weight) %>%
        summarize(mean_bf=mean(bayesfactor),
                  sd_bf=sd(bayesfactor),
                  .groups="drop")
    res
}
S <- c(100, 1000, seq(10e3, 50e3, by=10000))
results <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list)
results2 <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list,
            prior.weight=1)
combined <- bind_rows(results, results2)
```


```{r standardev, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    ggplot(aes(montecarlo.samples, sd_bf)) +
    geom_point(aes(color=prior.weight)) +
    geom_line(aes(group=prior.weight, color=prior.weight)) +
    scale_y_log10() +
    theme_bw(base_size=16) +
    xlab("Monte Carlo samples") +
    ylab("Standard deviation of\n log Bayes Factor")
```

Note that with importance sampling, relatively stable estimates for the Bayes factor are obtained with as few as 10,000 Monte Carlo samples while sampling from the prior distribution is very unstable for small sample sizes.

```{r means, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    filter(montecarlo.samples > 100) %>%
    ggplot(aes(prior.weight, mean_bf)) +
    geom_point() +
    theme_bw(base_size=16) +
    ylab("Mean log Bayes factor") +
    xlab("Prior weight")
```

## Relationship between priors, marginal likelihood, and Bayes factors

As previous discussed, the target distribution for importance sampling, denoted by *g*, is a mixture of the posterior and prior with a default weight of 0.1 on the prior.  The following code can be used to visualize the relationship between the posterior, prior, and *g* for mutations in cfDNA conditional on the tumor-derived model:




# Application to van't Erve et al.


```{r vanterve}
data(crcseq)
```

# Session information

```{r session}
sessionInfo()
```
