---
title: "Modeling the origin of mutations identified in a liquid biopsy: cancer or clonal hematopoiesis?"
author: "Adith S. Arun and Robert B. Scharpf"
date: "`r format(Sys.Date())`"
output: BiocStyle::html_document
abstract: >
  Mutation-based approaches for detection of cancer from cell-free DNA (cfDNA) using liquid biopsies have the potential to track a patient's response to treatment, enabling effective and timely decisions on therapy.  However, mutations arising from clonal hematopoeisis (CH) are common and tumor biopsies for definitive identification of the origin of these mutations is not always available. Sequencing of matched cells from buffy coat and the presence of mutations in these cells has been used as a test to rule-in CH, but uneven sequencing depths between matched cell-free DNA and buffy coat samples and the potential for contamination of buffy coat with circulating tumor cells (CTCs) are not captured by rule-based analyses.  This package provides estimates of the marginal likelihood for CH- and tumor-origin models of cfDNA mutations in plasma, requiring sequencing of cfDNA and matched buffy coat and the allele frequencies of high quality alignments available from standard mutation callers.
vignette: >
  %\VignetteIndexEntry{Modeling the origin of mutations in a liquid biopsy: cancer or clonal hematopoiesis?}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Motivating example

Suppose our plasma-based cfDNA analysis identified four fragments with a *TP53* variant out of 1000 fragments overlapping that position.  
In matched WBC sequencing, we observed no fragments with this mutation with only 600 fragments spanning this position.
How strong is the evidence that the *TP53* mutation is tumor-derived?
The possible scenarios that explain the observed data are:
(1) the variant is tumor-derived,
(2) the variant is derived from CH but we were unlucky and did not observe the mutation in the buffy coat.
A third explanation is sequencing error, but after removing PCR duplicates we assume the probability of observing the same mutation in four independent molecules is negligible.

# Data organization

```{r setup, message=FALSE}
library(magrittr)
library(tidyverse)
library(ggrepel)
library(plasmut)
library(qpdf)
knitr::opts_chunk$set(cache=TRUE)
lo <- function(p) log(p/(1-p))
```

We assume the following minimal representation for the mutation data such that each row uniquely identifies a mutation within a sample.

```{r data_wrangling}
## sample data
p53 <- tibble(y=c(4, 0),
              n=c(1000, 600),
              analyte=c("plasma", "buffy coat"),
              mutation="TP53",
              sample_id="id1")
dat <- p53 %>%
    unite("uid", c(sample_id, mutation), remove=FALSE) %>%
    group_by(uid) %>%
    nest()
## required format for plasmut
dat
```


# Approach and implementation

## Bayesian model

Let ($y_p$, $n_p$) denote the number of mutant reads and total number of distinct reads plasma and ($y_w$, $n_w$) the corresponding frequencies from the buffy coat. 
Assuming that the mutation is either tumor- or CH-derived, we can estimate the posterior odds:
$$\frac{p(S | y_p, y_w, n_p, n_w)}{p(H | y_p, y_w, n_p, n_w)} = \frac{p(y_w, y_p | n_p, n_w, S)}{p(y_w, y_p | n_p, n_w, H)} \cdot \frac{p(S)}{p(H)}, $$
where *S* is the somatic (tumor-derived) model and H is hematopietic (CH-derived).
The term $\frac{p(y_w, y_p | n_p, n_w, S)}{p(y_w, y_p | n_p, n_w, H)}$ is the Bayes factor and is a ratio of the marginal likelihoods.
For the denominator, we assume that the unobserved true mutant allele frequency ($\theta$) in plasma is the same as the mutant allele frequency in buffy coat and rewrite the marginal likelihood as $\int_{\theta} p(y_p | \theta, n_p) p(y_w | \theta,  n_w)d\theta p(\theta | H)$.
We suggest a diffuse prior for $\theta$ to provide support both rare and common CH mutations. 

For model *S*, the marginal likelihood is factored as $p(y_p, y_w | n_p, n_w, S) = \int_{\theta_p} p(y_p | n_p, \theta_p, S) d\theta_w\int_{\theta_w} p(y_w | n_w, \theta_w, S) d\theta_p$.
A diffuse prior for $\theta_p$ allows support for both rare and common somatic mutations. 
$\theta_W$ can be $> 0$ under model $S$ if circulating tumor cells (CTCs) are inter-mixed with white blood cells in the buffy coat.
As CTCs tend to be uncommon even in patients with late-stage disease, we use a prior distribution that concentrates most of the mass near zero (i.e., Beta(1, $10^3$)).

## Implementation

We can compute marginal likelihoods for model S and H by simply drawing Monte Carlo samples from the $\theta$ priors we defined. 
But, when the likelihood is more concentrated than the prior, this sampling approach can be inefficient and numerically unstable. 
To solve these issues, we implement an importance sampling approach. 
The general idea for importance sampling is that we want to sample $\theta$'s from a density that has a similar shape as the posterior but fatter tails so that we avoid spending time on simulations that have very low likelihood. Importance sampling should make estimation of the marginal likelihood more efficient, and the resulting Bayes factor estimates more stable with fewer Monte Carlo samples.

## Model assumptions and priors
Below, we assume that
(1) the probability that CTCs are mixed in with the WBCs is small (e.g., 1 CTC per 1,000 cells) using a Beta(1, 10^3) prior,
(2) the prior for the MAF of ctDNA in plasma is relatively diffuse (Beta(1, 9)), and
(3) the prior for germline or CHIP variants in WBCs is also relatively diffuse (Beta(1, 9)).

Our importance sampler will be a mix of the prior and posterior with a default weight of 0.1 on the prior, resulting in a fatter tailed distribution than the posterior.

```{r parameters}
## Parameters
param.list <- list(ctc=list(a=1, b=10e3),
                   ctdna=list(a=1, b=9),
                   chip=list(a=1, b=9),
                   montecarlo.samples=50e3,
                   prior.weight=0.1)
```

## Monte Carlo estimation of the marginal likelihood and Bayes factors

Below, we estimate the marginal likelihood for the mutation frequencies under two models:  (1) a somatic model that assumes that any mutations in plasma and buffy coat are tumor derived, and (2) a germline or CHIP model that assumes that the mutations observed in plasma and buffy coat are not tumor-derived.  All Monte Carlo samples are saved by default.  For running this model on large datasets, we recommend saving only the marginal likelihoods and Bayes factors by setting  `save_montecarlo=FALSE`.

```{r montecarlo}
stats <- importance_sampler(dat$data[[1]], param.list)
stats  ## with all the monte carlo samples
## Just the mutation-level summary statistics (marginal likelihoods and bayes factors)
importance_sampler(dat$data[[1]], param.list, save_montecarlo=FALSE)
```

With our prior that CTCs are rare, we view the plasma MAF of 4/1000 and buffy coat MAF of 1/1000 as very weak evidence that the mutation is tumor derived (log Bayes factor `round(stats$bayesfactor$bayesfactor, 2)`).  Had we instead observed no mutations in buffy coat, the log Bayes factor increases by 3 and we would conclude that there is strong evidence in support of a tumor-derived mutation.

```{r scenariob}
importance_sampler(dat$data[[2]], param.list, save_montecarlo=FALSE)
```

Note, that the log marginal likelihood of the CTC model is higher as observing no mutations in 1000 reads from buffy coat is more consistent with our prior; and the log marginal likelihood of the chip model decreases as we would anticipate observing mutant reads in the buffy coat if the mutation was derived from CHIP.

## Efficiency of importance sampler

As long as `montecarlo.samples` is big enough, we should obtain a similar estimate of the marginal likelihood without importance sampling.
Since our target distribution $g$ is a mixture of the prior and posterior with weight `prior.weight`, setting `prior.weight=1` just samples $\theta$'s from our prior (i.e., importance sampling is not implemented).  Below, we compare the stability of the Bayes factor estimate as a function of the Monte Carlo sample size and prior weight:

```{r prior.weight, cache=TRUE}
fun <- function(montecarlo.samples, data,
                param.list, prior.weight=0.1){
    param.list$montecarlo.samples <- montecarlo.samples
    param.list$prior.weight <- prior.weight
    res <- importance_sampler(data, param.list,
                              save_montecarlo=FALSE) %>%
        mutate(montecarlo.samples=montecarlo.samples,
               prior.weight=param.list$prior.weight)
    res
}
fun2 <- function(montecarlo.samples, data,
                 param.list, prior.weight=0.1,
                 nreps=100){
    res <- replicate(nreps, fun(montecarlo.samples, data,
                                param.list,
                                prior.weight=prior.weight),
                     simplify=FALSE) %>%
        do.call(bind_rows, .) %>%
        group_by(montecarlo.samples, prior.weight) %>%
        summarize(mean_bf=mean(bayesfactor),
                  sd_bf=sd(bayesfactor),
                  .groups="drop")
    res
}
S <- c(100, 1000, seq(10e3, 50e3, by=10000))
results <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list)
results2 <- S %>%
    map_dfr(fun2, data=dat$data[[1]], param.list=param.list,
            prior.weight=1)
combined <- bind_rows(results, results2)
```


```{r standardev, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    ggplot(aes(montecarlo.samples, sd_bf)) +
    geom_point(aes(color=prior.weight)) +
    geom_line(aes(group=prior.weight, color=prior.weight)) +
    scale_y_log10() +
    theme_bw(base_size=16) +
    xlab("Monte Carlo samples") +
    ylab("Standard deviation of\n log Bayes Factor")
```

Note that with importance sampling, relatively stable estimates for the Bayes factor are obtained with as few as 10,000 Monte Carlo samples while sampling from the prior distribution is very unstable for small sample sizes.

```{r means, fig.width=8, fig.height=5}
combined %>%
    mutate(prior.weight=factor(prior.weight)) %>%
    filter(montecarlo.samples > 100) %>%
    ggplot(aes(prior.weight, mean_bf)) +
    geom_point() +
    theme_bw(base_size=16) +
    ylab("Mean log Bayes factor") +
    xlab("Prior weight")
```


# Application to van't Erve et al.

We apply our methods to the dataset introduced in [van't Erve et al](https://aacrjournals.org/clincancerres/article/29/5/899/716607/Metastatic-Colorectal-Cancer-Treatment-Response). The cell-free DNA and buffy coat were sequenced for patients with metastatic colorectal cancer and we can estimate Bayes factors for these mutations. 

We select four mutations and run the importance sampler for these candidate mutations independently. 

```{r vanterve}
data(crcseq)

crcseq %>% select(-position)
```

```{r vanterve importance sampling}
params <- list(ctdna = list(a = 1, b = 9), 
               ctc = list(a = 1, b = 10^3), 
               chip = list(a= 1, b = 9), 
               montecarlo.samples = 50e3, 
               prior.weight = 0.1)

muts <- unite(crcseq, "uid", c(patient, gene), remove = FALSE) %>% 
        group_by(uid) %>% nest()

#Each element of the data column contains a table with the variant and total allele counts in plasma and buffy coat. 

#Run the importance sampler
muts$IS <- muts$data %>% map(importance_sampler, params)

bf <- apply(muts, 1, function(x) x$data %>% select(-position) %>% mutate(bayes_factor = x$IS$bayesfactor$bayesfactor)) 

bf %>% do.call(rbind, .) %>% as_tibble() %>% select(patient, gene, aa, bayes_factor) %>% distinct()
```

We report the log Bayes factor for the four mutations. The Bayes factor informs us of the likelihood that a mutation is tumor-derived. The log Bayes factor is equal to the log posterior odds when the prior odds of a mutation being tumor-derived are 1, which is a good uninformative assumption in the majority of cases. Let us examine the E1306* APC mutation for patient 12. When sequencing the buffy coat, we observe 0 variant reads across 963 total reads. But, when sequencing the cell-free DNA, we observed 395 variant reads across 1750 total reads. Intuitively, this means that the mutation is not found in white blood cells but is present in cell-free DNA fragments meaning that the source of the mutation must be from a non-WBC source. We expect the mutation to be a product of the tumor whose cellular turnover shed mutant DNA fragments into the bloodstream. Concordantly, the log Bayes factor is 190 which translates to a probability of being tumor derived of $\frac{exp(190)}{exp(190) + 1}$ or 1. 

On the other hand, we observe 5 mutant reads across 1495 total WBC reads for the M237K TP53 mutation; 15 mutant reads across 2969 total cfDNA reads. The observed mutant read rate is roughly equal in WBC and cfDNA, and we just happen to observe more cfDNA reads in general. This suggests that the variant is likely not tumor derived and probably results from CH. There is of course the possibility that there is an abnormally high CTC fraction in WBC harboring this mutation but this is not likely. We would expect the probability of the mutation being tumor derived to be low. As expected, this probability is estimated to be $\frac{exp(-1.1)}{exp(-1.1) + 1}$ or 0.25. 

# Session information

```{r session}
sessionInfo()
```


















